test/utils.py

```diff
@@ -217,6 +219,185 @@ def all_files_exist():
+        auto_confirm = kwargs.get('auto_confirm')
+        assert auto_confirm is None or input is None
+        assert input is None or isinstance(input, cls.Input)
+        return auto_confirm is not None or input.is_confirm
+
+    def __init__(self, tt_cmd, tmpdir, app_path, instances):
+        self.__app_name = os.path.basename(app_path)
+        self.__instances = instances
+        self.__tt_cmd = tt_cmd
+        self.__app_dir = shutil.copytree(app_path, os.path.join(tmpdir, self.__app_name))
+
+    def __run_tt(self, cmd, target, **kwargs):
+        cmd = [self.__tt_cmd, cmd]
+        input = kwargs.get('input')
+        auto_confirm = kwargs.get('auto_confirm')
+        assert input is None or auto_confirm is None
+        if auto_confirm is not None:
+            cmd.append(auto_confirm)
+        elif input is not None:
+            assert isinstance(input, self.Input)
+            input = input.str
+        if target is not None:
+            if target.inst is None:
+                cmd.append(self.__app_name)
+            else:
+                cmd.append(f'{self.__app_name}:{target.inst}')
+        return run_command_and_get_output(cmd, cwd=self.__app_dir, input=input)
+
+    def inst_id(self, inst_name):
+        return f"{self.__app_name}:{inst_name}"
+
+    @property
+    def app_dir(self):
+        return self.__app_dir
+
+    @property
+    def instances(self):
+        return self.__instances
+
+    def instances_of(self, *targets):
+        instances = []
+        for inst in self.instances:
+            if self.is_instance_of(inst, *targets):
+                instances.append(inst)
+        return instances
+
+    def status(self, target=app_target):
+        rc, out = self.__run_tt("status", target)
+        assert rc == 0
+        return extract_status(out)
+
+    def app_path(self, *paths):
+        return os.path.join(self.app_dir, *paths)
+
+    def run_path(self, inst, *paths):
+        return os.path.join(self.app_dir, run_path, inst, *paths)
+
+    def lib_path(self, inst, *paths):
+        return os.path.join(self.app_dir, lib_path, inst, *paths)
+
+    def log_path(self, inst, *paths):
+        return os.path.join(self.app_dir, log_path, inst, *paths)
+
+    def start(self, target=app_target):
+        return self.__run_tt("start", target)
+
+    def stop(self, target=app_target, **kwargs):
+        return self.__run_tt("stop", target, **kwargs)
+
+    def restart(self, target=app_target, **kwargs):
+        return self.__run_tt("restart", target, **kwargs)
+
+    def kill(self, target=app_target, **kwargs):
+        return self.__run_tt("kill", target, **kwargs)
+
+    def clean(self, target=app_target, **kwargs):
+        return self.__run_tt("clean", target, **kwargs)
+
+    def logrotate(self, target=app_target):
+        return self.__run_tt("logrotate", target)
+
+
+# Values to be used with pytest.mark.parametrize to check various input at confirmation prompt.
+tt_app_params_confirmation = [
```

Comment:
1) By the name of array is not clear what this array should do, and how it will be working. Seems a bit complex to support those tests.
2) `utils.py` does not seem the best place to keep tests arguments.

Reply:
1. Move it closer to where it's used. Hope along with comment it is more clear of how its working. It's just various values for one of test parameter that is passed to test via `pytest.mark.parametrize` machinery, so I see no problem to support those tests.
2. The same parameter set is used in multiple places, so initially I put it here just to avoid duplication. However, I'm agree it's bad place for it, so finally I just duplicate it where it's used.